File: src/main/java/com/polarbookshop/orderservice/book/BookClient.java
package com.polarbookshop.orderservice.book;

import java.time.Duration;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.client.WebClientResponseException;
import reactor.core.publisher.Mono;
import reactor.util.retry.Retry;

@Component
@RequiredArgsConstructor
public class BookClient {

  public static final String BOOKS_API = "/books/{isbn}";
  private final WebClient webClient;

  public Mono<Book> getBookByIsbn(String isbn) {
    return webClient
        .get()
        .uri(BOOKS_API, isbn)
        .retrieve()
        .bodyToMono(Book.class)
        .timeout(Duration.ofSeconds(3), Mono.empty())
        .onErrorResume(WebClientResponseException.class, e -> Mono.empty())
        .retryWhen(Retry.backoff(3, Duration.ofMillis(100)))
        .onErrorResume(Exception.class, e -> Mono.empty());
  }
}

File: src/main/kotlin/com/polarbookshop/orderservice/book/BookClient.kt
package com.polarbookshop.orderservice.book

import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.WebClientResponseException
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.onErrorResume
import reactor.util.retry.Retry
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds
import kotlin.time.toJavaDuration

private const val BOOKS_ROOT_API = "/books"

@Component
class BookClient(
    private val webClient: WebClient,
) {
    fun getBookByIsbn(isbn: String): Mono<Book> =
        webClient
            .get()
            .uri("$BOOKS_ROOT_API/{isbn}", isbn)
            .retrieve()
            .bodyToMono(Book::class.java)
            .timeout(3.seconds.toJavaDuration(), Mono.empty())
            .onErrorResume(WebClientResponseException.NotFound::class.java) { Mono.empty() }
            .retryWhen(Retry.backoff(3, 100.milliseconds.toJavaDuration()))
            .onErrorResume(Exception::class) { Mono.empty() }
}
